shapely/geoarrow.py

```diff
@@ -0,0 +1,163 @@
+    WKB = SchemaCalculator.ENCODING_WKB
+    WKT = SchemaCalculator.ENCODING_WKT
+    GEOARROW = SchemaCalculator.ENCODING_GEOARROW
+    GEOARROW_INTERLEAVED = SchemaCalculator.ENCODING_GEOARROW_INTERLEAVED
+
+
+class GeoArrowTypeExporter:
+
+    def __init__(self, encoding, geometry_type=None, dimensions=None):
+        encoding = Encoding(encoding).value
+        wkb_type = 0
+
+        if geometry_type == GeometryType.POINT:
+            wkb_type = 1
+        elif geometry_type in (GeometryType.LINESTRING, GeometryType.LINEARRING):
+            wkb_type = 2
+        elif geometry_type == GeometryType.POLYGON:
+            wkb_type = 3
+        elif geometry_type == GeometryType.MULTIPOINT:
+            wkb_type = 4
+        elif geometry_type == GeometryType.MULTILINESTRING:
+            wkb_type = 5
+        elif geometry_type == GeometryType.MULTIPOLYGON:
+            wkb_type = 6
+        elif geometry_type == GeometryType.GEOMETRYCOLLECTION:
+            wkb_type = 7
+        elif geometry_type is not None:
+            raise ValueError(f"Unknown value for geometry_type: '{geometry_type}'")
+
+        if dimensions is not None:
+            dimensions = dimensions.lower()
+
+        if dimensions == "xy" or dimensions is None:
+            wkb_type += 0
+        elif dimensions == "xyz":
+            wkb_type += 1000
+        elif dimensions == "xym":
+            wkb_type += 2000
+        elif dimensions == "xyzm":
+            wkb_type += 3000
+        else:
+            raise ValueError(f"Unknown value for dimensions: '{dimensions}'")
+
+        self._encoding = encoding
+        self._wkb_type = wkb_type
+
+    def __arrow_c_schema__(self):
+        return SchemaCalculator.from_wkb_type(self._encoding, self._wkb_type)
+
+
+class GeoArrowArrayExporter:
+
+    def __init__(self, geometries, *, preferred_encoding=None) -> None:
+        self._geometries = np.array(geometries)
+
+        if preferred_encoding is None:
+            self._preferred_encoding = Encoding.GEOARROW.value
+        else:
+            self._preferred_encoding = Encoding(preferred_encoding).value
+
+    def __arrow_c_schema__(self):
+        if self._preferred_encoding == Encoding.WKB.value:
+            return SchemaCalculator.from_wkb_type(self._preferred_encoding)
+
+        calculator = SchemaCalculator()
+        calculator.ingest_geometry(self._geometries)
+        return calculator.finish(self._preferred_encoding)
+
+    def __arrow_c_array__(self, requested_schema=None):
+        if requested_schema is None:
+            requested_schema = self.__arrow_c_schema__()
+
+        builder = ArrayBuilder(requested_schema)
+        builder.append(self._geometries)
+        chunks = builder.finish(ensure_non_empty=True)
+
+        if len(chunks) == 1:
+            return requested_schema, chunks[0]
+        else:
+            raise ValueError(
+                f"Can't export geometries to single chunk ({len(chunks)} required)"
+            )
```

Comment:
It's a little weird to me that we have both `__arrow_c_array__` and `__arrow_c_stream__` on the same class. I'd have usually expected that there would be a single underlying data representation in a class, and that would inform which PyCapsule API is defined on that class.

Maybe it makes sense to have both `ArrayBuilder` and `ChunkedArrayBuilder`? Maybe there are some occasions where a user wants to have all geometries in a single array. I.e. does `pyarrow.Table.from_pandas` always create a table with a single chunk? It looks like it does. In that case, exporting a non-chunked array of GeoArrow data would be important, so that the geometries could be appended on the table. And we'd presumably never want to raise an exception here.

(Ideally, I'd love `pyarrow.Table.from_pandas` to optionally export a chunked Table, but alas)

Reply:
> Functionally they are the same

Even in the cases where there's a single chunk under the hood, I disagree that they're functionally the same because usually there are distinct APIs for contiguous and chunked arrays. 

So far in my experience with the PyCapsule APIs, it's been expected that you can infer the storage format of the producer based on the dunder method that exists on the object. Is this not why `pyarrow.Array` and `pyarrow.RecordBatch` have `__arrow_c_array__` implemented but not `__arrow_c_stream__`, and vice versa for `pyarrow.ChunkedArray` and `pyarrow.Table`? 

I very often use the presence of the dunder methods to decide on the most efficient way to operate on data, and I'd much prefer community consensus around whether both APIs should ever exist on a single class.